### PR TITLE
Generate preamble code

### PR DESCRIPTION
**Pull request by ikehara**
I'd like to pass parameters to a contract as function arguments in rust code.
This might be helpful for client code generation because we can easily determine 
what parameters each contract entry point has by just parsing a source code. 

I'm not sure what kind of types should be accepted as parameters.
Do you think everything (convertible to bytes) should be ok or there should be 
a certain group of types which is acceptable for entry points.

### CODE REVIEW

File: hmcdk/codegen/src/lib.rs [hunk 1]
```diff
@@ -2,21 +2,81 @@ extern crate proc_macro;
...
+            let pair = type_path.path.segments.last().ok_or("internal error")?;
+            let ident = &pair.ident;
+            match types::to_abi_primitive(&ident.to_string()) {
+                Some(_p) => {
+                    let a = parse_quote! {
+                        let #name: #ident = get_arg(#index).unwrap();
+                        };
+                    Ok(a)
+                }
+                None => Err("invalid arg type".to_string())
```
Comment (bluele) [inline hmcdk/codegen/src/lib.rs]: please add type and argument info to error message

File: hmcdk/codegen/src/lib.rs [hunk 2]
```diff
@@ -2,21 +2,81 @@ extern crate proc_macro;
...
+            }
+        }
+        _ => Err("invalid arg type".to_string())
```
Comment (bluele) [inline hmcdk/codegen/src/lib.rs:18]: please add type and argument info to error message

File: hmcdk/abi/src/types.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+pub enum AbiInteger {
+    I8, I16, I32, I64,
+    U8, U16, U32, U64,
+}
+
+pub enum AbiPrimitive {
+    Integer(AbiInteger),
+    Bytes,
+    Address,
+    Hash,
```
Comment (bluele) [inline hmcdk/abi/src/types.rs]: Where does `Hash` comes from?

File: hmcdk/abi/src/types.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
+
```
Comment (bluele) [inline hmcdk/abi/src/types.rs]: do rustfmt

**Comment by bluele** [comment]
Thanks, @ikehara .

I'd like to accept all types which can convert into `Vec<u8>`, but do you think there is a problem at your work #64 ?

**Comment by ikehara** [comment]
> Thanks, @ikehara .
> 
> I'd like to accept all types which can convert into `Vec<u8>`, but do you think there is a problem at your work #64 ?

The code generation in #64 is done based on the parsed ast of a source file.
I think all unrecognized types should be mapped to "bytes" or something in the intermediate JSON file and converted into []byte in the go source code.

**Review by bluele — APPROVED** [review]
LGTM 👍

**Pull request merged**